### PR TITLE
Allow block children, remove aligns concern

### PR DESCRIPTION
**Pull request by Soreine**
Goals:

- support blocks in cells by default (including nested tables)
- keep the ability to have text children only
- remove aligns management. This will be implemented by users of the plugin, using new utils to ease the task.

For now, this PR is only 0.24.x compatible. I'll upgrade it to latest slate before merging.

### CODE REVIEW

**Comment by zhujinxuan** [comment]
Just want to mention that be careful about copy/paste in nested table when the table is nested with block.

This is an example happening in slate-deep-table.


I would like to submit a copy/paste PR if this slate-edit-table is released after two weeks later (when I have done and tested the copy/paste in my table solution)

**Comment by Soreine** [comment]
@jasonphillips any advice is welcome, since you paved the way for implementing tables containing blocks.

**Comment by Soreine** [comment]
So far, I haven't yet determined the best behavior for key handlers. Some of them could not be relevant anymore. For example, when should `ARROW_DOWN` move the cursor to the cell below it, if there are several blocks in the current cell?
I need to keep an easy way to customize these behaviors. For now I think it's fine, because you can always wrap the plugin's `onKeyDown` with your own and catch cases you're not interested in.

**Comment by zhujinxuan** [comment]
My personal suggestion is to use another table type for nested table;  When table is nested in table, then many behaviors are different.  Especially you do not want to have a deeply nested table by `getFragmentAtRange`, and `TablePosition` would be ugly in the child table of the parent table.  My current solution is to use another specialized table type with specialized utils/changes/other.

**Comment by jasonphillips** [comment]
It'll be good to have the option of moving back to the main library from my (at this point somewhat substantially diverged) fork at some point, so I look forward to trying this. I believe I may have a couple of opinionated differences in mine that won't apply here, but I'm not certain off the top of my head. 

Mine was built since supporting block-based tables represented a non-negotiable deal breaker for my users (faculty members) in order to fully move them off an addition to awful WYSIWYGs and completely into Slate.  

On the key presses -- were you planning to have a block-first approach (user presses Enter in a cell, new block / split block within that cell) or the opposite (Enter creates a new row as before, but shift+Enter breaks within a cell) ?

**Comment by Zhouzi** [comment]
Things not to forget:

- [ ] Update example and documentation (if need be)

File: package.json [hunk 1]
```diff
@@ -36,12 +37,14 @@
     "immutable": "^3.8.2",
     "mocha": "^3.0.1",
     "prettier": "^1.9.2",
+    "prop-types": "^15.6.0",
```
Comment (Zhouzi) [inline package.json]: I don't know why this dependency has been added, a mistake I guess.

**Comment by Zhouzi** [comment]
@Soreine This PR is now ready for review. I updated the code, docs, example and stuff to match the latest version of Slate. I don't have a deep understanding of the intention regarding alignment management though, is there anything more you wanted to add/remove on that side? I believe it's all good but maybe there's something more you wanted to do.

I played around with the example: navigation, controls, editing and everything is working as expected 👍

**Comment by zhujinxuan** [comment]
I find one bug and two problems existing in current plugin:
bug: backspace does not work in the paragraph head inside table cell.  This is an example of pressing backspace does not work:
!()[http://g.recordit.co/Y1g3lpZVJr.gif]

Two problems:
1. enter key insert a new row, rather than 
2. paste will be quite a headache 
!()[]

**Comment by Zhouzi** [comment]
Thank you very much for reporting @zhujinxuan 👍

I just created #71 #70 to address those issues. We're currently in a race to upgrade all of our code to Slate's latest version and since those issues are not affecting us, I am afraid we will not be able to provide a fix right now. We'd be really happy to review any PR though 👌

**Comment by zhujinxuan** [comment]
@Zhouzi Never mind~ Indeed, I prefer the old slate-edit-table (table_cell children as text instead of block) better.

**Comment by zhujinxuan** [comment]
@Zhouzi  BTW, how do you think about https://github.com/GitbookIO/slate-edit-list/pull/69?  It provides the module and source-map support for the consistency of the recent slate upgrade.

File: lib/validation/validateNode.js [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+/*
+ * Ensure each row has the same number of columns.
+ */
+function validateNode(opts: Options) {
+    return node => {
+        if (node.type !== opts.typeTable) {
+            return undefined;
+        }
+
+        const rows = node.nodes.filter(child => child.type === opts.typeRow);
```
Comment (Soreine) [inline lib/validation/validateNode.js]: Extract this part in a function named 'validateColumnCount'.
And see the previous implementation, it declares some short functions like `isCell`, `countColumn`, and uses `Range`. It was more readable.

File: lib/validation/validateNode.js [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+        if (rowsMissingColumns.isEmpty()) {
+            return undefined;
+        }
+
+        return change => {
+            rowsMissingColumns.forEach(row => {
+                const numberOfCellsToAdd = maxColumns - row.nodes.size;
+                Array.from({ length: numberOfCellsToAdd })
+                    .map(() => Block.create({ type: opts.typeCell }))
```
Comment (Soreine) [inline lib/validation/validateNode.js]: Use `createCell`

File: lib/validation/validateNode.js [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+            rows
+                .map(
+                    row =>
+                        row.nodes.filter(child => child.type === opts.typeCell)
+                            .size
+                )
+                .max()
+        );
+        const rowsMissingColumns = rows.filter(
+            row => row.nodes.size < maxColumns
```
Comment (Soreine) [inline lib/validation/validateNode.js]: I don't know in which order normalizations are applied, but `row.nodes.size` may represent the actual cell count if `row.nodes` does not only contain cells.

File: package.json [hunk 5]
```diff
@@ -1,48 +1,59 @@
 {
   "name": "slate-edit-table",
   "description": "A Slate plugin to handle keyboard events in tables.",
-  "version": "0.15.0-alpha.1",
+  "version": "0.14.3",
```
Comment (Soreine) [inline package.json]: Keep 0.15.0-alpha.1 until we release 0.15.0

**Comment by Soreine** [comment]
I can't request changes on my own PR

File: lib/validation/validateNode.js (hunk 3 above)
Comment (Soreine) [reply lib/validation/validateNode.js]: It handles default blocks in cell for you

**Comment by Soreine** [comment]
✅ Approved. Will merge later, once it was correctly integrated in gitbook

**Comment by CameronAckermanSEL** [comment]
Any reason this PR stops at only allowing one type of content block as opposed to multiple types?

**Comment by zhujinxuan** [comment]
@CameronAckermanSEL I did not see only allowing one contentBlock in `lib/validation/schema.js`

**Comment by CameronAckermanSEL** [comment]
@zhujinxuan the options for the plugin is `typeContent` singular as opposed to supporting an array of types.

**Comment by SamyPesse** [comment]
@CameronAckermanSEL `[typeContent: string]`: default type for blocks in cells. Also used as default type for blocks created when exiting the table with Mod+Enter.

It allows multiple types of block in cells. But the plugin need to know what is the default type when it has to create new cells.

**Comment by CameronAckermanSEL** [comment]
@SamyPesse, I see. I encountered this on a fork where I provided this feature by having the following interface for opts

```
{
   ...
   typeContent: []string,
   defaultType: string,
}
```
Alternatively, the default type could be discovered based on the first element in the `typeContent` array, whichever is more discoverable by consumers of the plugin.

**Comment by CameronAckermanSEL** [comment]
Wanted to also revisit this issue that was brought up in this PR https://github.com/GitbookIO/slate-edit-table/issues/71

You should be able to implement this feature by simply removing the existing onEnter handler and allowing slate to perform its default behavior.

**Pull request merged**